rfc/cdevents-translator.md

```diff
@@ -1,65 +1,225 @@
+### Goals and Non-Goals
+Goals:
+- Create shared packages/interfaces that will be used as common in translating/sending CDEvents
+- Using `cdevents-translator` library to implement Gerrit translator initially
 
-Example of initializing endpoints for Gerrit and GitHub in `cdevents-translator`, these endpoints needs to be configured in the respective SCM systems to receive events.
+Non-Goals:
+- Implement translators for other SCM tools or any application that needs a translation to CDEvents
 
+
+### Design
+Based on the initial design discussion we have come up with 3 different approaches to create `cdevents-translator` library, each one has it's own pros/cons to pick the best suitable approach.
+
+### Approach 1 : Creating a Web Service with Libraries/Packages for different translators
+A Golang main application "cdevents-translator" will be created to expose the translation functionality as a web service with HTTP/REST API server.</br>
+SCM system's webhooks/endpoints are registered to handle the translation of the events into CDEvents.</br>
+An interface `EventTranslator` will be created to handle `TranslateEvent` and this needs to be implemented by different translators.
+
+- Example structure of cdevents-translator application
 ````go
-POST("/gerrit-webhooks", gerrit.HandleTranslateGerritEvent)
-POST("/github-webhooks", github.HandleTranslateGitHubEvent)
+package main
+
+type EventTranslator interface {
+	TranslateEvent()
+}
+
+func main() {
+  
+   // Register translators in the main application
+   http.HandleFunc("/translate/gerrit-webhooks", gerrittranslator.TranslateEvent)
+   http.HandleFunc("/translate/github-webhooks", githubtranslator.TranslateEvent)
+
+   // Start the main application's HTTP server on port 8080
+   http.ListenAndServe(":8080", nil)
+}
+````
+#### Create a Library/Package gerrit-translator-cdevents
+Library with the name "gerrit-translator-cdevents" will be created in a separate package/repo.</br>
+Declare a struct to hold fields related to translation of Gerrit event.</br>
+Implement "TranslateEvent" to translate Gerrit event to CDEvent.
+
+- Example structure of gerrit-translator-cdevents library
+````go
+package gerrittranslator
+
+
+type GerritTranslator struct {
+   //Fields for Gerrit event Translator
+}
+
+func (translator *GerritTranslator) TranslateEvent(req *http.Request) (CDEvent, error) {
+	//handle translate Gerrit event to CDEvent
+}
 ````
-An interface `EventTranslator` will be created to handle `TranslateEvent`. 
-Different source of systems need to implement `TranslateEvent()` to handle translation from source events to CDEvents.
 
+#### Pros and Cons
+This structure can be extended by adding more translator libraries in the future without modifying the main application code.</br>
+The problem with this approach anyone who is wanting some custom set of translator would need to write their own main class.
+
+### Approach 2 : Creating a Go's plugin system to implement different Translators
```

Comment:
I'd probably rename this to "Using Go plugins to implement different Translators"